src/herder/TxSetUtils.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "util/asio.h"
+#include "TxSetUtils.h"
+#include "crypto/Hex.h"
+#include "crypto/Random.h"
+#include "crypto/SHA.h"
+#include "database/Database.h"
+#include "herder/SurgePricingUtils.h"
+#include "ledger/LedgerManager.h"
+#include "ledger/LedgerTxn.h"
+#include "ledger/LedgerTxnEntry.h"
+#include "ledger/LedgerTxnHeader.h"
+#include "main/Application.h"
+#include "main/Config.h"
+#include "transactions/TransactionUtils.h"
+#include "util/GlobalChecks.h"
+#include "util/Logging.h"
+#include "util/ProtocolVersion.h"
+#include "util/XDRCereal.h"
+#include "util/XDROperators.h"
+#include "xdrpp/marshal.h"
+
+#include <Tracy.hpp>
+#include <algorithm>
+#include <list>
+#include <numeric>
+
+namespace stellar
+{
+namespace TxSetUtils
+{
+
+UnorderedMap<AccountID, TxSetFrame::AccountTransactionQueue>
+buildAccountTxQueues(TxSetFrame const& txSet)
+{
+    ZoneScoped;
+    UnorderedMap<AccountID, TxSetFrame::AccountTransactionQueue> actTxQueueMap;
+    for (auto const& tx : txSet.getTransactions())
+    {
+        auto id = tx->getSourceID();
+        auto it = actTxQueueMap.find(id);
+        if (it == actTxQueueMap.end())
+        {
+            auto d = std::make_pair(id, TxSetFrame::AccountTransactionQueue{});
+            auto r = actTxQueueMap.insert(d);
+            it = r.first;
+        }
+        it->second.emplace_back(tx);
+    }
+
+    for (auto& am : actTxQueueMap)
+    {
+        // sort each in sequence number order
+        std::sort(am.second.begin(), am.second.end(),
+                  [](TransactionFrameBasePtr const& tx1,
+                     TransactionFrameBasePtr const& tx2) {
+                      return tx1->getSeqNum() < tx2->getSeqNum();
+                  });
+    }
+    return actTxQueueMap;
+}
+
+struct SurgeCompare
+{
+    Hash mSeed;
+    SurgeCompare() : mSeed(HashUtils::random())
+    {
+    }
+
+    // return true if tx1 < tx2
+    bool
+    operator()(TxSetFrame::AccountTransactionQueue const* tx1,
+               TxSetFrame::AccountTransactionQueue const* tx2) const
+    {
+        if (tx1 == nullptr || tx1->empty())
+        {
+            return tx2 ? !tx2->empty() : false;
+        }
+        if (tx2 == nullptr || tx2->empty())
+        {
+            return false;
+        }
+
+        auto& top1 = tx1->front();
+        auto& top2 = tx2->front();
+
+        auto cmp3 = feeRate3WayCompare(top1, top2);
+
+        if (cmp3 != 0)
+        {
+            return cmp3 < 0;
+        }
+        // use hash of transaction as a tie breaker
+        return lessThanXored(top1->getFullHash(), top2->getFullHash(), mSeed);
+    }
+};
+
+void
```

Comment:
This should return a new `TxSetFrame`. Mutation breaks, at least, the tx hash and likely some other invariants too.